api/client/container/utils.go

```diff
@@ -54,45 +57,59 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
 
 	eventChan := make(chan events.Message)
 	go eh.Watch(eventChan)
-	defer close(eventChan)
 
 	// Get events via Events API
 	f := filters.NewArgs()
 	f.Add("type", "container")
 	f.Add("container", containerID)
 	options := types.EventsOptions{
-		Since:   fmt.Sprintf("%d", since.Unix()),
 		Filters: f,
 	}
 	resBody, err := dockerCli.Client().Events(ctx, options)
 	if err != nil {
-		return -1, fmt.Errorf("can't get events from daemon: %v", err)
+		return nil, fmt.Errorf("can't get events from daemon: %v", err)
 	}
-	defer resBody.Close()
 
-	go system.DecodeEvents(resBody, func(event events.Message, err error) error {
-		if err != nil {
+	go func() {
+		eventsErr <- system.DecodeEvents(resBody, func(event events.Message, err error) error {
+			if err != nil {
+				return fmt.Errorf("decode events error: %v", err)
+			}
+			eventChan <- event
 			return nil
-		}
-		eventChan <- event
-		return nil
-	})
+		})
+		close(eventChan)
+	}()
 
-	if waitRemove {
-		select {
-		case <-destroyChan:
-			return exitCode, nil
-		case <-detachChan:
-			return 0, nil
+	go func() {
+		var waitErr error
+		if waitRemove {
+			select {
+			case <-destroyChan:
+				// keep exitcode and return
+			case <-detachChan:
+				atomic.StoreInt32(&exitCode, 0)
+			case waitErr = <-eventsErr:
+				atomic.StoreInt32(&exitCode, 125)
+			}
+		} else {
+			select {
+			case <-exitChan:
```

Comment:
Yep, good idea. Actually destroyChan, exitChan and detachChan can be combined into one.